cmd/client-nfd/subcmd/compat/validate-node.go

```diff
@@ -0,0 +1,125 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package compat
+
+import (
+	"context"
+	"fmt"
+	"os"
+	"strconv"
+	"time"
+
+	"github.com/jedib0t/go-pretty/v6/table"
+	"github.com/jedib0t/go-pretty/v6/text"
+	"github.com/spf13/cobra"
+	"oras.land/oras-go/v2/registry"
+
+	"sigs.k8s.io/node-feature-discovery/cmd/client-nfd/subcmd/compat/options"
+	pkgcompat "sigs.k8s.io/node-feature-discovery/pkg/client-nfd/compat"
+)
+
+var (
+	image     string
+	tags      []string
+	platform  options.PlatformOption
+	plainHTTP bool
+)
+
+var validateNodeCmd = &cobra.Command{
+	Use:   "validate-node",
+	Short: "Validate node based on image compatibility metadata",
+	Long:  "Validate node based on image compatibility metadata from the NFD artifact",
+	PreRunE: func(cmd *cobra.Command, args []string) error {
+		return platform.Parse(cmd)
+	},
+	RunE: func(cmd *cobra.Command, args []string) error {
+		ctx, cancel := context.WithTimeout(context.Background(), time.Minute)
+		defer cancel()
+
+		ref, err := registry.ParseReference(image)
+		if err != nil {
+			return err
+		}
+
+		out, err := pkgcompat.ValidateNode(ctx, &ref, tags, platform.Platform, plainHTTP)
+		if err != nil {
+			return err
+		}
+		pprint(out)
+
+		return nil
+	},
+}
+
+func pprint(css []*pkgcompat.CompatibilityStatus) {
+	for i, cs := range css {
+		fmt.Print(text.Colors{text.FgCyan, text.Bold}.Sprintf("Compatibility set #%d ", i+1))
+		fmt.Print(text.FgCyan.Sprintf("Weight: %d", cs.Weight))
+		if cs.Tag != "" {
+			fmt.Print(text.FgCyan.Sprintf("; Tag: %s", cs.Tag))
+		}
+		fmt.Println()
+		fmt.Println(text.FgWhite.Sprintf("Description: %s", cs.Description))
+
+		for _, r := range cs.Rules {
+			t := table.NewWriter()
+			t.SetStyle(table.StyleLight)
+			t.SetOutputMirror(os.Stdout)
+
+			validTxt := text.BgRed.Sprint(" fail ")
+			if r.IsMatch {
+				validTxt = text.BgGreen.Sprint(" ok ")
+			}
+			t.SetTitle(fmt.Sprintf("rule: %s - %s", text.Bold.Sprintf("%s", r.Name), validTxt))
+			t.Style().Title.Format = text.FormatUpper
+			t.AppendHeader(table.Row{"#", "feature", "expression", "status"})
+
+			for source, features := range r.MatchedFeatures {
+				for feature, elems := range features {
+					t.AppendSeparator()
+					for i, elem := range elems {
+
+						status := text.FgRed.Sprintf("%s", elem[pkgcompat.MatchKeyStatus])
+						if elem[pkgcompat.MatchKeyStatus] == pkgcompat.MatchStatusOK {
+							status = text.FgGreen.Sprintf("%s", elem[pkgcompat.MatchKeyStatus])
+						}
+
+						t.AppendRow(table.Row{strconv.Itoa(i + 1), fmt.Sprintf(
```

Comment:
lots of indent layers in a function, to make it easier to read, it's better to move some logic to another function. usually > 3 indent layers in one function will lead to difficult in reading the code.

Reply:
Fixed.